### PR TITLE
Yukon basic personal amount

### LINKED ISSUE #312: Yukon basic personal amount
reference:
https://www.canada.ca/content/dam/cra-arc/formspubs/pbg/td1yt-ws/td1yt-ws-23e.pdf

### PR DESCRIPTION
**Pull request by RuoqiTan**
fixes #312

### CODE REVIEW

File: policyengine_canada/variables/gov/provinces/yt/tax/income/credits/yt_basic_personal_amount.py
Comment (PavelMakarchuk) [inline policyengine_canada/variables/gov/provinces/yt/tax/income/credits/yt_basic_personal_amount.py:1]: this calculation is only applicable if the applicants income is more than $165,430 - if it is below the BPA is $13,521 flat.

Comment (PavelMakarchuk) [reply policyengine_canada/variables/gov/provinces/yt/tax/income/credits/yt_basic_personal_amount.py:1]: please include that clause

File: policyengine_canada/parameters/gov/provinces/yt/tax/income/credits/basic_personal_amount/income_threshold.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+description: Yukon uses this income threshold for the basic personal amount.
```
Comment (PavelMakarchuk) [inline policyengine_canada/parameters/gov/provinces/yt/tax/income/credits/basic_personal_amount/income_threshold.yaml:1]:
```suggestion
description: Yukon provides a calculation for the basic personal amount for filer with income above this threshold.
```

File: policyengine_canada/parameters/gov/provinces/yt/tax/income/credits/basic_personal_amount/base_amount.yaml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+description: Yukon uses this base amount for the basic personal amount.
```
Comment (PavelMakarchuk) [inline policyengine_canada/parameters/gov/provinces/yt/tax/income/credits/basic_personal_amount/base_amount.yaml:1]:
```suggestion
description: Yukon provides this base amount under the basic personal amount.
```

File: policyengine_canada/tests/gov/provinces/yt/tax/income/credits/yt_basic_personal_amount.yaml [hunk 3]
```diff
@@ -0,0 +1,23 @@
+- name: Not Eligible
+  period: 2023
+  input:
+    province_code: YT
+    individual_net_income: 160_000
+  output:
+    yt_basic_personal_amount: 0
```
Comment (PavelMakarchuk) [inline policyengine_canada/tests/gov/provinces/yt/tax/income/credits/yt_basic_personal_amount.yaml:7]:
```suggestion
    yt_basic_personal_amount: 13_521
```

File: policyengine_canada/parameters/gov/provinces/yt/tax/income/credits/basic_personal_amount/applicable_amount.yaml
Comment (PavelMakarchuk) [inline policyengine_canada/parameters/gov/provinces/yt/tax/income/credits/basic_personal_amount/applicable_amount.yaml:1]: could you please add the legal source

Comment (PavelMakarchuk) [reply policyengine_canada/parameters/gov/provinces/yt/tax/income/credits/basic_personal_amount/applicable_amount.yaml:1]: need to find the right page [here](https://laws.yukon.ca/cms/images/LEGISLATION/PRINCIPAL/2002/2002-0118/2002-0118.pdf)

File: policyengine_canada/parameters/gov/provinces/yt/tax/income/credits/basic_personal_amount/base_amount.yaml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+description: Yukon provides this base amount under the basic personal amount.
+values:
+  2023-01-01: 13_521
```
Comment (PavelMakarchuk) [inline policyengine_canada/parameters/gov/provinces/yt/tax/income/credits/basic_personal_amount/base_amount.yaml:3]:
```suggestion
  2023-01-01: 15_000
```

File: policyengine_canada/parameters/gov/provinces/yt/tax/income/credits/basic_personal_amount/base_amount.yaml [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+metadata:
+  unit: currency-CAD
+  label: Yukon basic personal amount base amount
+  reference:
+  - title: Worksheet for the Yukon 2023 Personal Tax Credits Return
+    href: https://www.canada.ca/content/dam/cra-arc/formspubs/pbg/td1yt-ws/td1yt-ws-23e.pdf
```
Comment (PavelMakarchuk) [inline policyengine_canada/parameters/gov/provinces/yt/tax/income/credits/basic_personal_amount/base_amount.yaml:9]: add [this](https://www.canada.ca/content/dam/cra-arc/formspubs/pbg/td1yt/td1yt-23e.pdf) source

File: policyengine_canada/parameters/gov/provinces/yt/tax/income/credits/basic_personal_amount/applicable_amount.yaml [hunk 6]
```diff
@@ -0,0 +1,11 @@
+description: Yukon uses this applicable amount for the basic personal amount.
+values:
+  2023-01-01: 1_479
+metadata:
+  unit: currency-CAD
+  label: Yukon basic personal amount applicable amount
+  reference:
+  - title: Worksheet for the Yukon 2023 Personal Tax Credits Return
+    href: https://www.canada.ca/content/dam/cra-arc/formspubs/pbg/td1yt-ws/td1yt-ws-23e.pdf
+  - title: Income Tax Act
```
Comment (PavelMakarchuk) [inline policyengine_canada/parameters/gov/provinces/yt/tax/income/credits/basic_personal_amount/applicable_amount.yaml:10]: add the exact part of the income tax act (paragraph, x...)

Comment (PavelMakarchuk) [reply policyengine_canada/parameters/gov/provinces/yt/tax/income/credits/basic_personal_amount/applicable_amount.yaml:10]: all yaml files

File: policyengine_canada/parameters/gov/provinces/yt/tax/income/credits/basic_personal_amount/applicable_amount.yaml [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+    href: https://laws.yukon.ca/cms/images/LEGISLATION/PRINCIPAL/2002/2002-0118/2002-0118.pdf
```
Comment (PavelMakarchuk) [inline policyengine_canada/parameters/gov/provinces/yt/tax/income/credits/basic_personal_amount/applicable_amount.yaml:11]: add page to link

Comment (PavelMakarchuk) [reply policyengine_canada/parameters/gov/provinces/yt/tax/income/credits/basic_personal_amount/applicable_amount.yaml:11]: all yaml files

File: policyengine_canada/parameters/gov/provinces/yt/tax/income/credits/basic_personal_amount/applicable_amount.yaml [hunk 8]
```diff
@@ -0,0 +1,11 @@
+description: Yukon uses this applicable amount for the basic personal amount.
```
Comment (PavelMakarchuk) [inline policyengine_canada/parameters/gov/provinces/yt/tax/income/credits/basic_personal_amount/applicable_amount.yaml:1]:
```suggestion
description: Yukon provides this applicable amount for the basic personal amount.
```

File: policyengine_canada/parameters/gov/provinces/yt/tax/income/credits/basic_personal_amount/applicable_amount.yaml
Comment (PavelMakarchuk) [inline policyengine_canada/parameters/gov/provinces/yt/tax/income/credits/basic_personal_amount/applicable_amount.yaml:1]: Could you make a folder called "reduction" and then put all the files in it that are not the base amount

File: policyengine_canada/variables/gov/provinces/yt/tax/income/credits/yt_basic_personal_amount.py [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+    defined_for = ProvinceCode.YT
+
+    def formula(person, period, parameters):
+        income = person("individual_net_income", period)
+        p = parameters(
+            period
+        ).gov.provinces.yt.tax.income.credits.basic_personal_amount
+        exceedance = income - p.income_threshold
+        eligible = exceedance >= 0
+        percent = max_(0, (p.scale_value - exceedance) / p.scale_value)
```
Comment (PavelMakarchuk) [inline policyengine_canada/variables/gov/provinces/yt/tax/income/credits/yt_basic_personal_amount.py:18]: break this formula up

File: policyengine_canada/variables/gov/provinces/yt/tax/income/credits/yt_basic_personal_amount.py [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+    label = "Yukon basic personal amount"
+    definition_period = YEAR
+    defined_for = ProvinceCode.YT
+
+    def formula(person, period, parameters):
+        income = person("individual_net_income", period)
+        p = parameters(
+            period
+        ).gov.provinces.yt.tax.income.credits.basic_personal_amount
+        exceedance = income - p.income_threshold
```
Comment (PavelMakarchuk) [inline policyengine_canada/variables/gov/provinces/yt/tax/income/credits/yt_basic_personal_amount.py:16]: if the person has income of under $165,430 they just get the main $13,521 - we need to account for that

File: policyengine_canada/parameters/gov/provinces/yt/tax/income/credits/basic_personal_amount/reduction/applicable_amount.yaml [hunk 11]
```diff
@@ -0,0 +1,14 @@
+description: Yukon provides this applicable amount for the basic personal amount.
+values:
+  2023-01-01: 1_479
+metadata:
+  unit: currency-CAD
+  label: Yukon basic personal amount applicable amount
+  reference:
+  - title: Worksheet for the Yukon 2023 Personal Tax Credits Return
+    href: https://www.canada.ca/content/dam/cra-arc/formspubs/pbg/td1yt-ws/td1yt-ws-23e.pdf
+  - title: Income Tax Act
```
Comment (PavelMakarchuk) [inline policyengine_canada/parameters/gov/provinces/yt/tax/income/credits/basic_personal_amount/reduction/applicable_amount.yaml:10]: add specific part of the legislation and add it to all the parameters

File: policyengine_canada/variables/gov/provinces/yt/tax/income/credits/yt_basic_personal_amount.py [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+        income = person("individual_net_income", period)
+        p = parameters(
+            period
+        ).gov.provinces.yt.tax.income.credits.basic_personal_amount
+        exceedance = income - p.reduction.income_threshold
+        eligible = exceedance >= 0
+        pct = (p.reduction.scale_value - exceedance) / p.reduction.scale_value
+        percent = max_(0, pct)
+        return (
+            (percent * p.reduction.applicable_amount) * eligible
```
Comment (PavelMakarchuk) [inline policyengine_canada/variables/gov/provinces/yt/tax/income/credits/yt_basic_personal_amount.py:21]: separate formula

**Review by MaxGhenis — CHANGES_REQUESTED** [review]
I'd suggest consolidating this into a single scale parameter which starts at `15_000` at income `0`, then phases out at a rate of `1_479 / 70_245` at income of `165_430`, then changes to a `0` rate at `165_430 + 70_245`

File: policyengine_canada/variables/gov/provinces/yt/tax/income/credits/yt_basic_personal_amount.py [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+
+class yt_basic_personal_amount(Variable):
+    value_type = float
+    entity = Person
+    label = "Yukon basic personal amount"
+    definition_period = YEAR
+    defined_for = ProvinceCode.YT
+
+    def formula(person, period, parameters):
+        income = person("individual_net_income", period)
```
Comment (PavelMakarchuk) [inline policyengine_canada/variables/gov/provinces/yt/tax/income/credits/yt_basic_personal_amount.py:12]: need to check for is_head

Comment (MaxGhenis) [reply policyengine_canada/variables/gov/provinces/yt/tax/income/credits/yt_basic_personal_amount.py:12]: no longer right

Comment (PavelMakarchuk) [reply policyengine_canada/variables/gov/provinces/yt/tax/income/credits/yt_basic_personal_amount.py:12]: No, we do not

File: policyengine_canada/variables/gov/provinces/yt/tax/income/credits/yt_basic_personal_amount.py [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+        head = person("is_head", period)
```
Comment (PavelMakarchuk) [inline policyengine_canada/variables/gov/provinces/yt/tax/income/credits/yt_basic_personal_amount.py:13]:
```suggestion
```

Comment (PavelMakarchuk) [reply policyengine_canada/variables/gov/provinces/yt/tax/income/credits/yt_basic_personal_amount.py:13]: my bad, we dont need to check that

File: policyengine_canada/variables/gov/provinces/yt/tax/income/credits/yt_basic_personal_amount.py [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+        p = parameters(
+            period
+        ).gov.provinces.yt.tax.income.credits.basic_personal_amount
+        reduction = p.reduction_rate.calc(income)
+        base = p.basic_personal_amount
+        return (base + reduction) * head
```
Comment (PavelMakarchuk) [inline policyengine_canada/variables/gov/provinces/yt/tax/income/credits/yt_basic_personal_amount.py:19]:
```suggestion
        return (base + reduction) 
```

File: policyengine_canada/parameters/gov/provinces/yt/tax/income/credits/basic_personal_amount/basic_personal_amount.yaml [hunk 16]
```diff
@@ -0,0 +1,9 @@
+description: Yukon provides this basic personal amount for people with net income less than income threshold.
+values:
+  2023-01-01: 15_000
+metadata:
+  unit: currency-CAD
+  label: Yukon basic personal amount for people with net income less than income threshold
+  reference:
+  - title: 2023 Yukon Personal Tax Credits Return
```
Comment (PavelMakarchuk) [inline policyengine_canada/parameters/gov/provinces/yt/tax/income/credits/basic_personal_amount/basic_personal_amount.yaml:8]: need Yukon income tax act reference

File: policyengine_canada/parameters/gov/provinces/yt/tax/income/credits/basic_personal_amount/basic_personal_amount.yaml [hunk 17]
```diff
@@ -0,0 +1,9 @@
+description: Yukon provides this basic personal amount for people with net income less than income threshold.
+values:
+  2023-01-01: 15_000
+metadata:
+  unit: currency-CAD
+  label: Yukon basic personal amount for people with net income less than income threshold
```
Comment (PavelMakarchuk) [inline policyengine_canada/parameters/gov/provinces/yt/tax/income/credits/basic_personal_amount/basic_personal_amount.yaml:6]:
```suggestion
  label: Yukon basic personal amount 
```

File: policyengine_canada/parameters/gov/provinces/yt/tax/income/credits/basic_personal_amount/basic_personal_amount.yaml [hunk 18]
```diff
@@ -0,0 +1,9 @@
+description: Yukon provides this basic personal amount for people with net income less than income threshold.
```
Comment (PavelMakarchuk) [inline policyengine_canada/parameters/gov/provinces/yt/tax/income/credits/basic_personal_amount/basic_personal_amount.yaml:1]:
```suggestion
description: Yukon credits this basic personal amount.
```

**Comment by PavelMakarchuk** [comment]
> 



> I'd suggest consolidating this into a single scale parameter which starts at `15_000` at income `0`, then phases out at a rate of `1_479 / 70_245` at income of `165_430`, then changes to a `0` rate at `165_430 + 70_245`

Not sure if I understood this correctly, but it seems like there has to be some logic implemented before we can support this right?

**Pull request closed without merging**